monai/data/dataset.py

```diff
@@ -93,22 +95,38 @@ def __init__(self, data, transform, cache_num=sys.maxsize, cache_rate=1.0):
                 will take the minimum of (cache_num, data_length x cache_rate, data_length).
             cache_rate (float): percentage of cached data in total, default is 1.0 (cache all).
                 will take the minimum of (cache_num, data_length x cache_rate, data_length).
+            num_workers (int): the number of worker processes to use.
+                If 0 a single thread will be used. Default is 0.
         """
         if not isinstance(transform, Compose):
             transform = Compose(transform)
         super().__init__(data, transform)
         self.cache_num = min(cache_num, int(len(self) * cache_rate), len(self))
         self._cache = list()
         print('Load and cache transformed data...')
-        for i in range(self.cache_num):
-            process_bar(i + 1, self.cache_num)
-            item = data[i]
-            for _transform in transform.transforms:
-                # execute all the deterministic transforms before the first random transform
-                if isinstance(_transform, Randomizable):
-                    break
-                item = apply_transform(_transform, item)
-            self._cache.append(item)
+        if num_workers > 0:
+            def _update_bar(i):
+                process_bar(i + 1, self.cache_num)
+            with Pool(num_workers) as p:
+                for i in range(self.cache_num):
+                    p.apply_async(self._load_cache_item, args=(i, transform, data), callback=_update_bar)
+                p.close()
+                p.join()
+        else:
+            for i in range(self.cache_num):
+                process_bar(i + 1, self.cache_num)
+                self._load_cache_item(i, transform, data)
+
+    def _load_cache_item(self, i, transform, data):
+        item = data[i]
+        for _transform in transform.transforms:
+            # execute all the deterministic transforms before the first random transform
+            if isinstance(_transform, Randomizable):
+                break
+            item = apply_transform(_transform, item)
+        self._cache.append(item)
```

Comment:
Tips: The CI test error is because you append value to `self._cache` in child process which will not update the `self._cache` in the main process. So even after running `_load_cache_item` for all items, the `self._cache` in main process is still empty.
Maybe you can try to return the transformed item and append it to `self._cache` in the main process.
Thanks.

Reply:
Yes you're right, thank you. I will update this and try with threading as you suggested.

Reply:
You are welcome.
Looking forward to your speed test result of `multi-process` vs `multi-thread`.